### PR TITLE
Fix FactoryVec state changes

### PR DESCRIPTION
**Pull request by V02460**
`FactoryVec` was acting weird and all over the place for me and it might be #84, too. I had some fun and modeled its fields as a state machine. Brings `push`, `pop`, and `clear` in line with the described model and fixes the indexing problems in the `generate` method. I think it is (more) correct now, but one could probably be smarter about `clear` and the use of `Vec::remove`.

### CODE REVIEW

**Review by AaronErhardt — APPROVED** [review]
Looks good! Thanks for the explanation in the code, that should make it a lot easier to understand.

Is it ready to merge?

**Comment by V02460** [comment]
I pushed one small fix. Now it’s fine from my side. Works for me™ :stuck_out_tongue:

In all seriousness, though… I should add some tests.

**Comment by AaronErhardt** [comment]
Yes, tests are definitely a good idea :)
It's hard to reason whether the code works or not.

**Comment by V02460** [comment]
Here are some unit tests 🎁 I made two separate commits, one for the tests themself and another one for getting GTK to play nicely along with multi-threaded test running. I changed the `Cargo.toml` for two reasons:
1. I got advised by the gtk-rs guys to adapt their test running helper, so I did that.
2. To use relm4 macros in the implementation of the tests.

File: Cargo.toml [hunk 1]
```diff
@@ -24,7 +24,9 @@ members = [
 adw = { version = "0.1", optional = true, package = "libadwaita" }
 async-trait = { version = "0.1", optional = true }
 fragile = "1.1"
+futures-channel = "0.3"
 futures-core = "0.3"
+futures-executor = "0.3"
```
Comment (AaronErhardt) [inline Cargo.toml:29]: Shouldn't those also be dev-depencencies?

**Review by AaronErhardt — COMMENTED** [review]
Looks good. Actually, some of this testing infrastructure could eventually be exposed to help resolving https://github.com/AaronErhardt/Relm4/issues/78. But that's another topic.

File: Cargo.toml (hunk 1 above)
Comment (V02460) [reply Cargo.toml]: Yes, pushed the change

**Comment by AaronErhardt** [comment]
Thanks a lot!

**Pull request merged**